### PR TITLE
Minify asmjs module imports

### PR DESCRIPTION
**Pull request by juj**
This minifies JS function imports to asm.js, using same minification scheme as in https://github.com/WebAssembly/binaryen/pull/1719/files#diff-402f63d949f69ca3feb553dcda0a5f26R44 .

Also this PR contains the same test from #7414, adjusted to pass for this case, so closing that PR.

### CODE REVIEW

**Comment by juj** [comment]
Getting odd looking failures with

```
#!/bin/bash -eo pipefail
flake8
./emscripten.py:49:1: E302 expected 2 blank lines, found 1
./emscripten.py:56:1: E302 expected 2 blank lines, found 1
./emscripten.py:250:1: E302 expected 2 blank lines, found 1
./tests/test_other.py:9077:31: E701 multiple statements on one line (colon)
Exited with code 1
```

but I am not sure that there are and formatting errors on those lines?

File: emscripten.py [hunk 1]
```diff
@@ -306,14 +318,22 @@ def move_preasm(m):
...
 
-  bg_funcs = basic_funcs + global_funcs
-  bg_vars = basic_vars + global_vars
+  basic_funcs = define_asmjs_import_names(basic_funcs, shared.Settings.MINIFY_ASMJS_IMPORT_NAMES)
+  global_funcs = define_asmjs_import_names(global_funcs, shared.Settings.MINIFY_ASMJS_IMPORT_NAMES)
+  basic_vars = define_asmjs_import_names(basic_vars, shared.Settings.MINIFY_ASMJS_IMPORT_NAMES)
+  global_vars = define_asmjs_import_names(global_vars, shared.Settings.MINIFY_ASMJS_IMPORT_NAMES)
+
+  def join_dicts(a, b):
+    return dict(a, **b)
```
Comment (kripken) [inline emscripten.py]: maybe @sbc100 has thoughts on the python style here?

File: tests/test_other.py [hunk 2]
```diff
@@ -9069,3 +9069,25 @@ def test_link(self):
...
+  def test_js_function_imports_are_minified(self):
+    def check_size(f, expected_size):
+      if not os.path.isfile(f): return # Nonexistent file passes in this check
+      obtained_size = os.path.getsize(f)
+      print('size of generated ' + f + ': ' + str(obtained_size))
+      try_delete(f)
+      assert obtained_size < expected_size
+
+    run_process([PYTHON, path_from_root('tests', 'gen_many_js_functions.py'), 'library_long.js', 'main_long.c'])
+    for wasm in [['-s', 'WASM=0', '--separate-asm']]: # TODO: Add Wasm testing ['-s', 'WASM=1'] to this list
```
[the remark this answers is not in the document]
Comment (juj) [reply tests/test_other.py]: Sorry, not sure what you mean? That line does have the TODO about wasm. Let me rephrase the text to refer to the implementation about wasm (and not just to add a test)

File: tests/test_other.py [hunk 3]
```diff
@@ -9069,3 +9069,25 @@ def test_link(self):
...
+      for closure in [['--closure', '1']]: # TODO: Implement minification also to when closure is not used
```
[the remark this answers is not in the document]
Comment (juj) [reply tests/test_other.py]: No, I noticed while doing this that if one does not use Closure, we do not do good uglify based minification, so wanted to add a comment about that for later, for non-Closure using developers

**Comment by juj** [comment]
The test `test_iostream_and_determinism` fails here, but I am unable to reproduce it locally. Have you seen this error?

File: tests/test_other.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (kripken) [reply tests/test_other.py]: Sorry, I should have been clearer. I meant, remove the TODO and add that testing for wasm :)

File: tests/test_other.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kripken) [reply tests/test_other.py]: I don't follow, what goes wrong without closure? The names should still be minified. Or is this test checking something even stronger?

File: emscripten.py [hunk 4]
```diff
@@ -39,6 +40,20 @@
...
+
+
+# Converts list of imports ['foo', 'bar', ...] to a dictionary of
+# name mappings in form { 'minified': 'unminified', ... }
+def define_asmjs_import_names(imports, minify_names):
+  minified = {}
+  for name in imports:
+    import_name = minified_js_names.generate() if minify_names else name
+    minified[import_name] = name
+  return minified
```
Comment (sbc100) [inline emscripten.py]: Maybe move this down to where its used?  Could even be an inner function within `function_tables_and_exports`?

Also we can drop the second argument and just use shared.Settings.MINIFY_ASMJS_IMPORT_NAMES directly?

Also what do you think of:

```
if shared.Settings.MINIFY_ASMJS_IMPORT_NAMES:
    return {minified_js_names.generate(): i for i in imports}
else:
    return {i: i for i in imports}
```

**Comment by kripken** [comment]
About that error, I'd guess that this PR breaks build determinism somehow - testing locally with different numbers of cores, and running the test many many times, has helped me debug such things in the past.

One guess - are those python operations guaranteed to be deterministic in both python 2 and 3 (maybe the iteration order on the dictionaries can change)?

File: emscripten.py (hunk 1 above)
Comment (sbc100) [reply emscripten.py]: i'm not familiar with that syntax.   Its not clear how that is working.   I assume its the same as 

````
rtn = dict(a)
rtn.update(b)
return rtn
````

Is there any change the keys can collide here?

File: tests/gen_many_js_functions.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
+import sys
+
+NUM_FUNCS_TO_GENERATE = 1000
+
+def func_name(i):
+	return 'thisIsAFunctionWithVeryLongFunctionNameThatWouldBeGreatToBeMinifiedWhenImportingToAsmJsOrWasmSideCodeToCallOtherwiseCodeSizesWillBeLargeAndNetworkTransfersBecomeVerySlowThatUsersWillGoAwayAndVisitSomeOtherSiteInsteadAndThenWebAssemblyDeveloperIsSadOrEvenWorseNobodyNoticesButInternetPipesWillGetMoreCongestedWhichContributesToGlobalWarmingAndThenEveryoneElseWillBeSadAsWellEspeciallyThePolarBearsAndPenguinsJustThinkAboutThePenguins' + str(i+1)
```
Comment (sbc100) [inline tests/gen_many_js_functions.py]: Emscripten uses 2 space indentation for python files.

File: tests/gen_many_js_functions.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
+import sys
```
Comment (sbc100) [inline tests/gen_many_js_functions.py]: Add copyright header:

```
# Copyright 2018 The Emscripten Authors.  All rights reserved.                   
# Emscripten is available under two separate licenses, the MIT license and the   
# University of Illinois/NCSA Open Source License.  Both these licenses can be   
# found in the LICENSE file.  
```

Comment (sbc100) [reply tests/gen_many_js_functions.py]: And `#!/usr/bin/env python` as line 0

File: tests/gen_many_js_functions.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+NUM_FUNCS_TO_GENERATE = 1000
+
+def func_name(i):
+	return 'thisIsAFunctionWithVeryLongFunctionNameThatWouldBeGreatToBeMinifiedWhenImportingToAsmJsOrWasmSideCodeToCallOtherwiseCodeSizesWillBeLargeAndNetworkTransfersBecomeVerySlowThatUsersWillGoAwayAndVisitSomeOtherSiteInsteadAndThenWebAssemblyDeveloperIsSadOrEvenWorseNobodyNoticesButInternetPipesWillGetMoreCongestedWhichContributesToGlobalWarmingAndThenEveryoneElseWillBeSadAsWellEspeciallyThePolarBearsAndPenguinsJustThinkAboutThePenguins' + str(i+1)
+
+def generate_js_library_with_lots_of_functions(out_file):
+	f = open(out_file, 'w')
```
Comment (sbc100) [inline tests/gen_many_js_functions.py]: Maybe use `with open(out_file, 'w') as f:` and then drop the `close()`, here and below

File: tests/test_other.py [hunk 8]
```diff
@@ -9069,3 +9069,26 @@ def test_link(self):
...
+      if not os.path.isfile(f):
+        return # Nonexistent file passes in this check
+      obtained_size = os.path.getsize(f)
+      print('size of generated ' + f + ': ' + str(obtained_size))
+      try_delete(f)
+      assert obtained_size < expected_size
+
+    run_process([PYTHON, path_from_root('tests', 'gen_many_js_functions.py'), 'library_long.js', 'main_long.c'])
+    for wasm in [['-s', 'WASM=0', '--separate-asm']]: # TODO: Add support to Wasm to minify imports, and then add Wasm testing ['-s', 'WASM=1'] to this list
+      for closure in [['--closure', '1']]: # Currently we rely on Closure for good minification of the JS function names. TODO: Add minification also for non-Closure users and add [] to this list to test minification without Closure.
```
Comment (sbc100) [inline tests/test_other.py]: Please put comments on the preceding line and wrap at 80.

File: tools/minified_js_name_generator.py [hunk 9]
```diff
@@ -0,0 +1,42 @@
+class MinifiedJsNameGenerator(object):
```
Comment (sbc100) [inline tools/minified_js_name_generator.py]: Add copyright header, and a quick comment about what this is used for?

File: tests/test_other.py [hunk 10]
```diff
@@ -9069,3 +9069,26 @@ def test_link(self):
...
+        args = [PYTHON, EMCC, '-O3', '--js-library', 'library_long.js', 'main_long.c', '-o', 'a.html'] + wasm + closure
+        print(' '.join(args))
+        run_process(args)
+        check_size('a.js', 80000)
+        check_size('a.wasm', 80000)
+        check_size('a.asm.js', 80000)
+    try_delete('library_long.js')
+    try_delete('main_long.c')
```
Comment (sbc100) [inline tests/test_other.py]: No need to clean these up, the test framework should cleanup anything files you created in CWD.

File: tests/gen_many_js_functions.py (hunk 5 above)
Comment (sbc100) [reply tests/gen_many_js_functions.py]: It not clear to me why the test relies on long symbol names, or a large number of symbols.   Doesn't minification effect all symbols, no matter how long, or how numerous?   Wouldn't a much small handwritten test be fine?

File: tests/gen_many_js_functions.py [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+	for i in range(NUM_FUNCS_TO_GENERATE):
+		f.write('int ' + func_name(i) + '(void);\n')
+
+	f.write('\nint main() {\n')
+	f.write('  int sum = 0;\n')
+
+	for i in range(NUM_FUNCS_TO_GENERATE):
+		f.write('  sum += ' + func_name(i) + '();\n')
+
+	f.write('\n  printf("Sum of numbers from 1 to ' + str(NUM_FUNCS_TO_GENERATE) + ': %d (expected ' + str((NUM_FUNCS_TO_GENERATE * (NUM_FUNCS_TO_GENERATE+1))/2) + ')\\n", sum);\n');
```
Comment (sbc100) [inline tests/gen_many_js_functions.py]: I was expecting to see this code get run somewhere, but it looks like the test just compiles it never runs it.  Is that intentional?

**Comment by sbc100** [comment]
> About that error, I'd guess that this PR breaks build determinism somehow - testing locally with different numbers of cores, and running the test many many times, has helped me debug such things in the past.
> 
> One guess - are those python operations guaranteed to be deterministic in both python 2 and 3 (maybe the iteration order on the dictionaries can change)?

Ah,  maybe you want OrderedDict() rather than plain old dict?

File: tests/test_other.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (juj) [reply tests/test_other.py]: Oh gotcha - I thought the minification support for wasm hadn't landed yet in incoming, now testing wasm part as well, it does pass, so added it in.

File: tests/test_other.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (juj) [reply tests/test_other.py]: Without closure, not all references to the long function names get minified. The JS side names are still in the long form, so left a TODO to highlight that if someone wants to look at it later.

File: tests/gen_many_js_functions.py (hunk 5 above)
Comment (juj) [reply tests/gen_many_js_functions.py]: Large number of functions with long names is so that it will be obvious when minification occurs, i.e. one can just look at plain generated size of the program to see if minification happened or not.

File: tests/test_other.py (hunk 8 above)
Comment (juj) [reply tests/test_other.py]: Previously tailing comments have been fine, can flake warn about these if that is now a mandated style thing?

File: tests/test_other.py (hunk 10 above)
Comment (juj) [reply tests/test_other.py]: It sounds like this kind of change has made `EMTEST_DETECT_TEMPFILE_LEAKS` check be a nop, which is supposed to verify when running tests that Emscripten does not leak test files behind. If we automatically clean all test files, we cannot test for leaks anymore.

Deleted the two lines, not sure if `EMTEST_DETECT_TEMPFILE_LEAKS` is no longer a thing.

File: tests/gen_many_js_functions.py (hunk 11 above)
Comment (juj) [reply tests/gen_many_js_functions.py]: Thanks, added a run.

File: tests/test_other.py [hunk 12]
```diff
@@ -9079,3 +9079,30 @@ def test_link(self):
...
+        return # Nonexistent file passes in this check
+      obtained_size = os.path.getsize(f)
+      print('size of generated ' + f + ': ' + str(obtained_size))
+      try_delete(f)
+      assert obtained_size < expected_size
+
+    run_process([PYTHON, path_from_root('tests', 'gen_many_js_functions.py'), 'library_long.js', 'main_long.c'])
+    # TODO: Add support to Wasm to minify imports, and then add Wasm testing ['-s', 'WASM=1'] to this list
+    for wasm in [['-s', 'WASM=1'], ['-s', 'WASM=0']]:
+      # Currently we rely on Closure for good minification of the JS function names.
```
Comment (kripken) [inline tests/test_other.py]: Maybe change the text here to "full minification of every single JS function name"?

Also, the function name `test_js_function_imports_are_minified` may be misleading as you are checking something even stronger?

Comment (juj) [reply tests/test_other.py]: Sounds good, adjusted.

**Comment by juj** [comment]
Even with OrderedDict seems to give errors. Rebased to incoming to trigger rerun, though running locally, it does pass on `binaryen*.test_iostream_and_determinism`, trying with EMCC_CORES=1 and without.

File: tests/test_other.py (hunk 8 above)
Comment (sbc100) [reply tests/test_other.py]: Its not strictly part of style guide I guess, but in cases like this where it makes the line longer than 80 I think it makes the code hard to read.

Ideally we'd be able to enforce some kind of line length limit, but its hard because (8) flake8 is not configurable in this way (b) there is so much code with long lines in emscripten already.

There was a recent discussion about limiting line length to 80 for new code, but we didn't decide to do that yet, and it would take a little effort to enforce programatically.

Anyway, its not a big deal here.

File: tests/test_other.py (hunk 10 above)
Comment (sbc100) [reply tests/test_other.py]: EMTEST_DETECT_TEMPFILE_LEAKS doesn't include changes you make in your own directory.

tearDown does this before it runs the checks:

```
       os.chdir(os.path.join(self.get_dir(), '..'))
       try_delete(self.get_dir())
```

Seems like its always worked that way.

I think this makes sense, we are more interested in stuff that is left in TMPDIR, and asking all tests to clean up after themselves adds a lot of noise the test code.
